classes/metric/concurrent_users_metrics.php

```diff
@@ -0,0 +1,134 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\metric;
+
+use cltr_database\collector;
+
+/**
+ * Metric class for new users.
+ *
+ * @package    metric_foobar
+ * @author     Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @copyright  2022, Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class concurrent_users_metrics extends builtin {
+    /**
+     * The metric's name.
+     *
+     * @return string
+     */
+    public function get_name(): string {
+        return 'concurrentusers';
+    }
+
+    /**
+     * The metric type.
+     *
+     * @return int
+     */
+    public function get_type(): int {
+        return manager::TYPE_GAUGE;
+    }
+
+    /**
+     * Retrieves the metric.
+     *
+     * @param  $backfilledrecord If data is backfilled, record is passed here.
+     * @return metric_item
+     */
+    public function get_metric_item($record = null): metric_item {
+        return new metric_item($this->get_name(), $record->time, $record->value, $this);
+    }
+
+    /**
+     * Metric's ability to be backfilled.
+     *
+     * @return bool
+     */
+    public function is_backfillable(): bool {
+        return true;
+    }
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param array $data Array of data specifying timeframe and sample to backfill.
+     * @return array|null
+     */
+    public function get_backfilled_metric(array $data): ?array {
+        global $DB;
+        $collector = new collector();
```

Comment:
This is concerning. Metrics are not supposed to be aware of collectors. Especially specific ones. You should be able to match any collector to any metric and have it work.

Reply:
I have split and renamed this function, please let me know if this suits.